pennylane/device_tracker.py

```diff
@@ -0,0 +1,171 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+from collections import defaultdict
+
+import pennylane as qml
+
+
+def track(dev, version="default", **kwargs):
+    r"""Creates a tracking context and applies it to a device.
+
+    Args:
+        dev (~.Device): a PennyLane-compatible device
+        version (str): name of tracker to use.  The current options are
+            `default` and `timing`.
+
+    Keyword Args:
+        reset_on_enter=True (bool): whether or not to reset information
+            entering the context
+
+    **Usage Information**
+
+    Note that with backpropagation, this functions should take ``qnode.device``
+    instead of the device used to create the QNode.
+
+    .. code-block:: python
+
+        dev = qml.device('default.qubit', wires=1)
+
+        @qml.qnode(dev, diff_method="parameter-shift")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+    With the default version, total execution information is printed on
+    each device execution.  The printed data depends on the device and tracker version,
+    but for standard PennyLane devices, the object will track executions and shots.
+
+    >>> with qml.track(circuit.device) as tracker:
+    ...    qml.grad(circuit)(0.1, shots=10)
+    Total: executions = 1	shots = 10
+    Total: executions = 2	shots = 20
+    Total: executions = 3	shots = 30
+
```

Comment:
I really like being able to query the `tracker` after a workflow and seeing the data, but I have to admit I wasn't sure on the usefulness of printing during execution. Isn't it just going to print:
```Total: executions = N        shots = 10 * N```
Also, if `shots` is not specified, then it just prints:
```
Total: executions = 1	
Total: executions = 2	
Total: executions = 3	
Total: executions = 4	
Total: executions = 5	
```

Reply:
Also, what is the implication of changing the `version`?  Is it just to determine what is printed?

It seems that querying `timing_tracker.history` still lets you see the number of shots, so isn't the `timing` version better / encompassing the default version (unless there's a reason not to track timing?).

To me it seems like it would be better to:
- Not have to specify a version
- Not have the printing and just be able to access info post-workflow
- If a user wants to print mid-workflow, can they print `tracker.shots` or `tracker.num_executions`?

Happy to discuss though as I'm joining the discussion a bit late on this one.

Reply:
I was inspired by my experience submitting optimizations workflows on Braket.  Each gradient involves many circuit executions, and you can't print any information out till the entire optimization step is done.  This way, you get information more frequently. This will also work well in conjunction with shot-adaptive optimizers since the number of shots will vary over time.

I designed the default tracker to be as flexible as possible, making no assumptions about the information the device might pass in.  This way, it's fairly easy for devices to pass in extra information to track as well instead of just execution number and shots.  For example, braket can add price.

I also want it to be as easy to adapt as possible.  A user should be easily able to create an inheriting class with a different `record` or `update` function.  For example, I intend to develop one that sends information to a logger instead of to `print`. Or maybe someone could modify `update` to also store information about current RAM usage.  Or we could have a tracker that just stores the data but doesn't print anything out.

Reply:
Maybe there could be a boolean flag, something like `print_while_tracking` or `with_print`? :thinking: